langchain/llms/sagemaker_endpoint.py

```diff
@@ -0,0 +1,132 @@
+"""Wrapper around Sagemaker InvokeEndpoint API."""
+import json
+from typing import Any, List, Mapping, Optional
+
+from pydantic import BaseModel, Extra
+
+from langchain.llms.base import LLM
+from langchain.llms.utils import enforce_stop_tokens
+
+VALID_TASKS = ("text2text-generation", "text-generation")
+
+
+class SagemakerEndpoint(LLM, BaseModel):
+    """Wrapper around custom Sagemaker Inference Endpoints.
+
+    To use, you should pass the AWS IAM Role and Role Session Name as
+    named parameters to the constructor.
+
+    Only supports `text-generation` and `text2text-generation` for now.
+    """
+
+    """
+    Example:
+        .. code-block:: python
+
+            from langchain import SagemakerEndpoint
+            endpoint_name = (
+                "https://runtime.sagemaker.us-west-2.amazonaws.com/endpoints/abcdefghijklmnop/invocations"
+            )
+            se = SagemakerEndpoint(
+                endpoint_name=endpoint_name,
+                role_arn="role_arn",
+                role_session_name="role_session_name"
+            )
+    """
+
+    endpoint_name: str = ""
+    """# The name of the endpoint. Must be unique within an AWS Region."""
+    task: Optional[str] = None
+    """Task to call the model with. Should be a task that returns `generated_text`."""
+    model_kwargs: Optional[dict] = None
+    """Key word arguments to pass to the model."""
+
+    role_arn: Optional[str] = None
+    role_session_name: Optional[str] = None
+
+    class Config:
+        """Configuration for this pydantic object."""
+
+        extra = Extra.forbid
+
+    @property
+    def _identifying_params(self) -> Mapping[str, Any]:
+        """Get the identifying parameters."""
+        _model_kwargs = self.model_kwargs or {}
+        return {
+            **{"endpoint_name": self.endpoint_name, "task": self.task},
+            **{"model_kwargs": _model_kwargs},
+        }
+
+    @property
+    def _llm_type(self) -> str:
+        """Return type of llm."""
+        return "sagemaker_endpoint"
+
+    def _call(self, prompt: str, stop: Optional[List[str]] = None) -> str:
+        """Call out to Sagemaker inference endpoint.
+
+        Args:
+            prompt: The prompt to pass into the model.
+            stop: Optional list of stop words to use when generating.
+
+        Returns:
+            The string generated by the model.
+
+        Example:
+            .. code-block:: python
+
+                response = se("Tell me a joke.")
+        """
+        import boto3
+
+        session = boto3.Session(profile_name="test-profile-name")
+        sagemaker_runtime = session.client("sagemaker-runtime", region_name="us-west-2")
```

Comment:
SageMaker client is being created for each request. This is not ideal, as it does not allow for connection reuse and general application performance. It might be better to move the session and the client creation to the class initialization block if possible.